141 - Datetime granularity

DeclanSL opened the pull request:
### Description
Added granularity for datetime fields. datetime will now default to millisecond granularity, but this behaviour can be changed via the profile.
The following units are acceptable: (millis, seconds, minutes, hours, days, months, years).
An example would be: { "field": "foo", "is": "granularTo", "value": "seconds" }

### Changes
- Added Timescale enum, and replaced the usage of chronounit with this. 
- Added AtomicConstraint IsGranularToDateConstraint
- Renamed AtomicConstraint IsGranularToConstraint to IsGranularToNumericConstraint
- Updated FieldSpecFactory to include IsGranularToDateConstraint and IsGranularToNumericConstraint.
- Updated SequentialDateIterator, RandomDateIterator,DateTimeFieldValueSource so we can generate dates to a given granularity.
- Updated AtomicConstraintReaderLookup so that AtomicConstraintType.ISGRANULARTO is handled for both datetime and decimal fields
- Updated DateTimeRestrictionsMerger class to take granularity in to account, the most coarse granularity restriction is kept.
- Added Cucumber tests : GranularTo-DateTime.feature
- Updated Cucumber tests GranularTo-Decimal.feature
- Added SequentialDateIterator tests (SequentialDateIteratorTest.java)
- Added Timescale Tests (TimescaleTest.java)
- Updated JsonProfileReaderTests.java, ProfileViolationTests.java and DateTimeRestrictionsMergerTests.java to include tests of granularity.
- Updated datahelix.schema.json to allow (millis, seconds, minutes, hours, days, months, years) values with the IsGranularTo constraint.


### Issue
Resolves #141

r-stuart commented on generator/src/main/java/com/scottlogic/deg/generator/generation/fieldvaluesources/datetime/DateTimeFieldValueSource.java:
```diff
@@ -5,6 +5,7 @@
 import com.scottlogic.deg.generator.utils.FilteringIterator;
 import com.scottlogic.deg.generator.utils.RandomNumberGenerator;
 import com.scottlogic.deg.generator.utils.UpCastingIterator;
+import org.jetbrains.annotations.Contract;
```
Unused import

DeclanSL replied:
I agree, I will remove this.

r-stuart commented on generator/src/main/java/com/scottlogic/deg/generator/generation/fieldvaluesources/datetime/RandomDateIterator.java:
```diff
@@ -25,11 +27,26 @@ public boolean hasNext() {
+    }
+
+    private int nanoToMilli(int nano) {
+        int factor = 1_000;
```
`1_000_000` for nano to milli?

DeclanSL replied:
I agree with this, and have resolved the issue.

r-stuart commented on generator/src/main/java/com/scottlogic/deg/generator/generation/fieldvaluesources/datetime/RandomDateIterator.java:
```diff
@@ -25,11 +27,26 @@ public boolean hasNext() {
+
+    private int nanoToMilli(int nano) {
+        int factor = 1_000;
+        return (nano / factor) * 1_000;
```
`(nano  / factor) * factor`

DeclanSL replied:
Agree, and have resolved this.

r-stuart commented on generator/src/main/java/com/scottlogic/deg/generator/generation/fieldvaluesources/datetime/Timescale.java:
```diff
@@ -0,0 +1,82 @@
+    }
+
+    private static int nanoToMilli(int nano) {
+        int factor = 1_000;
```
DRY between this and RandomDateIterator. I think it belongs here. Also it should reference the enum field explicitly.

DeclanSL replied:
Have removed this from the RandomDateIterator. it was unused in that class

r-stuart commented on generator/src/main/java/com/scottlogic/deg/generator/generation/fieldvaluesources/datetime/Timescale.java:
```diff
@@ -0,0 +1,82 @@
+        return Arrays.stream(Timescale.values())
+            .filter(t -> t.name.equals(name))
+            .findFirst()
+            .orElseThrow(() -> new IllegalArgumentException("Must be one of the supported datetime units (millis, seconds, minutes, hours, days, months, years"));
```
Generate this list of names from the enum rather than duplicating them.

DeclanSL replied:
Good idea, I have made the change.

r-stuart commented on generator/src/test/java/com/scottlogic/deg/generator/restrictions/DateTimeRestrictionsMergerTests.java:
```diff
@@ -1,21 +1,36 @@
+import com.scottlogic.deg.generator.generation.fieldvaluesources.datetime.Timescale;
+import org.assertj.core.api.AutoCloseableSoftAssertions;
 import org.junit.Assert;
+import org.junit.Before;
```
Unused import

DeclanSL replied:
agree, I've removed this

pdaulbyscottlogic commented:
apologies for the long comment.

If you add and run this test, it will fail because the lower bounds will not be inclusive, when they should be.

```
import static com.shazam.shazamcrest.MatcherAssert.assertThat;
import static com.shazam.shazamcrest.matcher.Matchers.sameBeanAs;

    @Test
    void merge_minOfDifferentGranularity_shouldReturnMostCoarse() {
        // ARRANGE
        DateTimeRestrictions left = new DateTimeRestrictions(Timescale.HOURS) {{
            min = new DateTimeLimit(REFERENCE_TIME, true); }};

        DateTimeRestrictions right = new DateTimeRestrictions(Timescale.MILLIS) {{
            min = new DateTimeLimit(REFERENCE_TIME.plusSeconds(1), false); }};

        // ACT
        MergeResult<DateTimeRestrictions> result = dateTimeRestrictionsMerger.merge(left, right);

        // ASSERT
        DateTimeRestrictions expecteddt = new DateTimeRestrictions(Timescale.HOURS);
        expecteddt.min = new DateTimeLimit(REFERENCE_TIME.plusHours(1), true);
        MergeResult<DateTimeRestrictions> expected = new MergeResult<>(expecteddt);

        assertThat(result, sameBeanAs(expected));
    }
```

DeclanSL commented:
> apologies for the long comment.
> 
> If you add and run this test, it will fail because the lower bounds will not be inclusive, when they should be.
> 
> ```
> import static com.shazam.shazamcrest.MatcherAssert.assertThat;
> import static com.shazam.shazamcrest.matcher.Matchers.sameBeanAs;
> 
>     @Test
>     void merge_minOfDifferentGranularity_shouldReturnMostCoarse() {
>         // ARRANGE
>         DateTimeRestrictions left = new DateTimeRestrictions(Timescale.HOURS) {{
>             min = new DateTimeLimit(REFERENCE_TIME, true); }};
> 
>         DateTimeRestrictions right = new DateTimeRestrictions(Timescale.MILLIS) {{
>             min = new DateTimeLimit(REFERENCE_TIME.plusSeconds(1), false); }};
> 
>         // ACT
>         MergeResult<DateTimeRestrictions> result = dateTimeRestrictionsMerger.merge(left, right);
> 
>         // ASSERT
>         DateTimeRestrictions expecteddt = new DateTimeRestrictions(Timescale.HOURS);
>         expecteddt.min = new DateTimeLimit(REFERENCE_TIME.plusHours(1), true);
>         MergeResult<DateTimeRestrictions> expected = new MergeResult<>(expecteddt);
> 
>         assertThat(result, sameBeanAs(expected));
>     }
> ```
> (i have not looked at the rest of the PR yet)

I have added the test to cover this scenario, and believe I have fixed the issue.

MattCline-SL commented on generator/src/main/java/com/scottlogic/deg/generator/restrictions/DateTimeRestrictionsMerger.java:
```diff
@@ -43,17 +78,35 @@
-                    return left;
-                return right;
+                if (left.getLimit().compareTo(right.getLimit()) > 0){
+                    return new DateTimeRestrictions.DateTimeLimit(left.getLimit(), (left.isInclusive() || right.isInclusive()) || inclusiveOverride);
```
`left.isInclusive() || right.isInclusive()) || inclusiveOverride` is repeated at least 3 times, can we make this a well named variable instead?

MattCline-SL commented on generator/src/main/java/com/scottlogic/deg/generator/restrictions/ParsedDateGranularity.java:
```diff
@@ -0,0 +1,27 @@
+
+/**
+ * Granularity expressions could be interpreted differently depending on other constraints on a field (eg, type constraints),
+ * so we store all possible parsings in this class, ready to make a GranularityRestrictions object
```
Do we? I can't see GranularityRestrictions anywhere else. Looks like an irrelevant comment for an old implementation. Rewrite this or delete it.

MattCline-SL commented on generator/src/main/java/com/scottlogic/deg/generator/restrictions/ParsedDateGranularity.java:
```diff
@@ -0,0 +1,27 @@
+import com.scottlogic.deg.generator.utils.NumberUtils;
+
+import java.math.BigDecimal;
+import java.sql.Time;
```
Where do we use `sql.Time`?

sl-slaing replied:
Same for `BigDecimal`

MattCline-SL commented on generator/src/test/java/com/scottlogic/deg/generator/cucumber/features/operators/GranularTo-DateTime.feature:
```diff
@@ -0,0 +1,50 @@
+
+  Scenario Outline: User is able to specify supported temporal granularities
+    Given foo is granular to <unit>
+    And foo is after 2000-01-01T00:00:00.000Z
```
Can we add some "before" tests too?

d-withers replied:
Tests for other constraints now added

MattCline-SL commented on generator/src/test/java/com/scottlogic/deg/generator/generation/fieldvaluesources/datetime/SequentialDateIteratorTest.java:
```diff
@@ -0,0 +1,123 @@
+package com.scottlogic.deg.generator.generation.fieldvaluesources.datetime;
```
There's a lot of useless whitespace in here? Autoformat or shorten it a bit

MattCline-SL commented on generator/src/test/java/com/scottlogic/deg/generator/generation/fieldvaluesources/datetime/SequentialDateIteratorTest.java:
```diff
@@ -0,0 +1,123 @@
+    }
+
+    @Test
+    void checkHasNextWhenCurrentIsEqualToMax() {
```
I think we do have a bit of a standard with unit tests. Arrange/Act/Assert and using methodName_withInputs_doesSomething as a test name.

sl-slaing commented on generator/pom.xml:
```diff
@@ -150,7 +150,16 @@
+            <scope>test</scope>
         </dependency>
+        <!-- https://mvnrepository.com/artifact/org.assertj/assertj-core -->
+        <dependency>
```
What does this dependency introduce into the project?
What is the licence for the dependency?

DeclanSL replied:
assertj allows fluent syntax to be used to write assertions when unit testing. @r-stuart has used it previously, and has used it here to write some of his unit tests.
See: https://github.com/joel-costigliola/assertj-core

The licence is Apache 2.0 see: https://github.com/joel-costigliola/assertj-core/blob/master/LICENSE.md

DeclanSL replied:
Dependency removed

sl-slaing commented on generator/src/main/java/com/scottlogic/deg/generator/restrictions/DateTimeRestrictionsMerger.java:
```diff
@@ -43,17 +78,35 @@
+                if (left.getLimit().compareTo(right.getLimit()) > 0){
+                    return new DateTimeRestrictions.DateTimeLimit(left.getLimit(), (left.isInclusive() || right.isInclusive()) || inclusiveOverride);
+                }
+                else
```
`else` isn't required

sl-slaing commented on generator/src/main/java/com/scottlogic/deg/generator/restrictions/DateTimeRestrictionsMerger.java:
```diff
@@ -43,17 +78,35 @@
+                if (left.getLimit().compareTo(right.getLimit()) < 0){
+                    return new DateTimeRestrictions.DateTimeLimit(left.getLimit(), (left.isInclusive() || right.isInclusive() ) || inclusiveOverride);
+                }
+                  else{
```
`else` isn't required

sl-slaing commented on generator/src/test/java/com/scottlogic/deg/generator/generation/fieldvaluesources/datetime/TimescaleTest.java:
```diff
@@ -0,0 +1,46 @@
+
+import static org.junit.jupiter.api.Assertions.*;
+
+class TimescaleTest {
```
Tests should be pluralised, `TimescaleTests`

DeclanSL replied:
Renamed TimescaleTest to TimescaleTests

sl-slaing commented on generator/src/test/java/com/scottlogic/deg/generator/generation/fieldvaluesources/datetime/TimescaleTest.java:
```diff
@@ -0,0 +1,46 @@
+
+class TimescaleTest {
+
+    public OffsetDateTime granularity(Timescale timescale, OffsetDateTime t) {
```
Should be private, probably static also?

DeclanSL replied:
Method removed, as it was unused.

sl-slaing commented on generator/src/test/java/com/scottlogic/deg/generator/generation/fieldvaluesources/datetime/TimescaleTest.java:
```diff
@@ -0,0 +1,46 @@
+    public void millisGranularity() {
+        OffsetDateTime imprecise = OffsetDateTime.of(2001, 1, 1, 1, 1, 1, 1, ZoneOffset.UTC);
+
+        Timescale.MILLIS.getGranularityFunction().apply(imprecise);
```
What are we asserting here?

sl-slaing commented on generator/src/test/java/com/scottlogic/deg/generator/restrictions/DateTimeRestrictionsTests.java:
```diff
@@ -85,56 +97,56 @@ public void matchShouldReturnFalseIfGivenSomethingThatIsntADateTime(){
 
         boolean result = restrictions.match(123);
 
-        Assert.assertFalse(result);
```
Is there a reason to remove `Assert.`?

DeclanSL replied:
I have put them back.

sl-slaing commented on schemas/src/main/resources/profileschema/0.1/datahelix.schema.json:
```diff
@@ -350,7 +350,14 @@
                   1E-19,
-                  1E-20
+                  1E-20,
+                  "millis",
```
Are these the correct case; are these required to be lower case?

DeclanSL replied:
Yes the supported datetime units must be provided in lowercase:
Value is not accepted.
Valid values: 1, 0.1, 0.01, 0.001, 0.0001, 0.00001, 0.000001, 1e-7, 1e-8, 1e-9, 1e-10, 1e-11, 1e-12, 1e-13, 1e-14, 1e-15, 1e-16, 1e-17, 1e-18, 1e-19, 1e-20, "millis", "seconds", "minutes", "hours", "days", "months", "years".

We followed the ofType convention here:
Valid values: "decimal", "integer", "string", "datetime".

DeclanSL commented on generator/src/main/java/com/scottlogic/deg/generator/generation/fieldvaluesources/datetime/DateTimeFieldValueSource.java:
```diff
@@ -50,14 +50,14 @@ public long getValueCount() {
 
-            if (granularity == ChronoUnit.MILLIS) return (duration.getNano() / 1000000) +1;
+            if (granularity == Timescale.MILLIS) return (duration.getNano() / 1000000) + 1;
             // future work to add customisable datetime granularity #141
```
Comment removed

DeclanSL commented on generator/src/main/java/com/scottlogic/deg/generator/generation/fieldvaluesources/datetime/Timescale.java:
```diff
@@ -0,0 +1,92 @@
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
+
+/**
```
NANOs wasn't part of the acceptance criteria in the feature request. 
Do we now want NANOs?

DeclanSL commented on generator/src/main/java/com/scottlogic/deg/generator/restrictions/DateTimeRestrictionsMerger.java:
```diff
@@ -1,5 +1,10 @@
 package com.scottlogic.deg.generator.restrictions;
```
I have auto formatted this file now.

DeclanSL commented on generator/src/main/java/com/scottlogic/deg/generator/constraints/atomic/IsGranularToDateConstraint.java:
```diff
@@ -0,0 +1,69 @@
+    public final ParsedDateGranularity granularity;
+
+    public IsGranularToDateConstraint(Field field, ParsedDateGranularity granularity, Set<RuleInformation> rules) {
+        this.granularity = granularity;
```
null checks added to IsGranularToDateConstraint and IsGranularToNumericConstraint

DeclanSL commented on generator/src/main/java/com/scottlogic/deg/generator/restrictions/DateTimeRestrictionsMerger.java:
```diff
@@ -29,10 +35,39 @@
+
+
+            default:
+                throw new UnsupportedOperationException();
```
message added

DeclanSL commented on generator/src/main/java/com/scottlogic/deg/generator/inputs/AtomicConstraintReaderLookup.java:
```diff
@@ -56,155 +58,179 @@
                         ParsedGranularity.parse(getValidatedValue(dto, Number.class)),
-                        rules));
+                        rules);
+                } catch (InvalidProfileException | IllegalArgumentException e) {
```
I have implemented some tryParse methods, and now only attempt to create a constraint when appropriate. I don't think it is possible to completely avoid using try{} catch{} here as we need to deserialize the profile input, before we know how to treat it.

pdaulbyscottlogic commented on generator/src/main/java/com/scottlogic/deg/generator/inputs/AtomicConstraintReaderLookup.java:
```diff
@@ -56,155 +58,179 @@
+                } catch (InvalidProfileException | IllegalArgumentException e) {
+                }
+
+                throw new InvalidProfileException(String.format("Field [%s]: Couldn't recognise granularity value, it must be either a negative power of ten or one of the supported datetime units.", dto.field));
```
this throw should be part of the catch statement block

DeclanSL commented on generator/src/main/java/com/scottlogic/deg/generator/restrictions/DateTimeRestrictions.java:
```diff
@@ -1,40 +1,65 @@
 public class DateTimeRestrictions {
+    private static final Timescale DEFAULT_GRANULARITY = Timescale.MILLIS;
+    private final Timescale granularity;
     public DateTimeLimit min;
```
Would require a change in about 50 places, as the min, max variable are set in other areas. I would think its probably worth doing it in another PR

MattCline-SL commented on generator/src/test/java/com/scottlogic/deg/generator/generation/fieldvaluesources/datetime/SequentialDateIteratorTest.java:
```diff
@@ -0,0 +1,124 @@
+    }
+
+    @Test
+    void HasNext_WhenCurrentIsEqualToMax_IsCorrect() {
```
Small point about method naming, should start with lowercase e.g. `hasNext_whenCurrentIsEqualToMax_isCorrect`

MattCline-SL commented on generator/src/test/java/com/scottlogic/deg/generator/generation/fieldvaluesources/datetime/SequentialDateIteratorTest.java:
```diff
@@ -0,0 +1,124 @@
+import java.time.OffsetDateTime;
+import java.time.ZoneOffset;
+
+class SequentialDateIteratorTest {
```
I feel like our test classes are normally public but I'm not sure?
